### DISCUSSION TITLE
Font size control on Wayland

### DISCUSSION THREAD

**Discussion by xofyarg** [wezterm/wezterm#4754] [General]
There are two common terms that related to font size: DPI and scale factor. As font size is often measured by points, some sources suggest to adjust DPI to make font size consistent across monitors. From what I know, however, DPI is a physical property of a monitor, and program is able to obtain it on Wayland. That means the only user configurable option should be the scale factor. So the pixel size could be calculated from point size and DPI, and then scaled by the scale factor into the final geometry.

I tried to apply my understanding into wezterm's config, and found that the DPI is somehow fixed: 192 on HiDPI screen or 96 otherwise, and I'm not sure if I'm supposed to tweak this? Or I misunderstood some of the concepts. Any advises are welcome.

P.S. With fractional scaling supported, for each monitor, I adjusted scale factor based on the density and distance. For some program, I specify font by pixel size if possible, so the size of both UI and font keep the same when moving between monitors.

**Comment by wez** [comment]
> and program is able to obtain it on Wayland

No, Wayland does not report the DPI to applications. It's value is not available to applications via the Wayland protocols.

Wezterm assumes a fixed DPI as a result: https://wezfurlong.org/wezterm/config/lua/config/dpi.html

The Wayland protocol only supports integer scaling.
If a compositor provides fractional scaling, that happens as an additional step in the compositor, invisible to the application.
An application can only see integer scale factor values.

**Comment by xofyarg** [comment]
Thanks @wez for the quick reply!

> Wayland does not report the DPI to applications.

Not directly, but can be calculated by physical size(`wl_output::geometry`) and resolution(`wl_output::mode`) of the display.

> The Wayland protocol only supports integer scaling.

The [fractional scale](https://wayland.app/protocols/fractional-scale-v1) [was merged](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/143) last February, and already in staging now.

> invisible to the application

I think that's the point, as not only for the font size, the UI needs to be scaled as well. So basically, the user has a reasonable application config, which is independent to the output/monitor. And the compositor is in charge of finding the right geometry to display(for both UI and font), by including the scale factor into consideration. Although I don't feel that's necessary, the application is able to acquire the scale factor via the above Wayland protocol.

**Reply by wez** [reply]
ah, well, that's good to hear that it is evolving in the right direction.
wezterm's wayland support could do with some attention, I'd welcome PRs from folks that know things about the protocol.
